File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.isbalanced = True

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self.dfs(root, 0)
        return self.isbalanced

    def dfs(self, root, depth):
        if root is None:
            return depth

        ldepth = self.dfs(root.left, depth + 1)
        rdepth = self.dfs(root.right, depth + 1)

        if abs(ldepth - rdepth) > 1:
            self.isbalanced = False
            return 0
        
        return max(ldepth, rdepth)
        
        
        

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {

public:
    int diameterOfBinaryTree(TreeNode* root) {
        int res = 0;
        diameterHelper(root, res);
        return res;
    }

    int diameterHelper(TreeNode* root, int& res) {
        if (root == nullptr) return 0;
        int left = diameterHelper(root->left, res);
        int right = diameterHelper(root->right, res);
        res = std::max(res, left + right);
        return 1 + std::max(left, right);
    }
};
